src/ScopeProvider.tsx

```diff
@@ -1,108 +1,188 @@
+       * @param target The `anotherAtom` that this atom is accessing.
+       * @returns The actual atom to access. If the atom is scoped, return an
+       * interceptedAtomCopy. Otherwise, return the unscoped original atom.
+       *
+       * Check the example below, when calling useAtomValue, jotai-scope will first
+       * find its intercepted copy (lets call it `anAtomIntercepted`). Then,
+       * `anAtom.read(get => get(dependencyAtom))` becomes
+       * `anAtomIntercepted.read(get => get(getAtom(anAtom, dependencyAtom)))`
+       * @example
+       * const anAtom = atom(get => get(dependencyAtom))
+       * const Component = () => {
+       *  useAtomValue(anAtom);
+       * }
+       * const App = () => {
+       *   return (
+       *    <ScopeProvider atoms={[anAtom]}>
+       *      <Component />
+       *    </ScopeProvider>
+       *   );
+       * }
+       */
+      const getAtom = <A extends AnyAtom>(orig: AnyAtom, target: A): A => {
+        // If a target is got/set by itself, then it is derived.
+        // Target could be an intercepted copy, so target is on the left.
+        if (isSelfAtom(target, orig)) {
+          // Since it is not derived, we check if it is marked as scoped.
+          return markedAsScoped
+            ? // If it is scoped, then current scope's intercepted copy is the store key
+              getInterceptedAtomCopy(target)
+            : // Otherwise, find the correct store key in the parent's scope.
+              // Check `getStoreKey` for details.
+              getParentStoreKey(target);
         }
-        return getScopedAtom(target);
+        // If a target is got/set by another atom, route the access to the
+        // target's intercepted copy, then repeat the procedure.
+        return getInterceptedAtomCopy(target);
       };
-      const scopedAtom: typeof anAtom = {
-        ...anAtom,
-        ...('read' in anAtom && {
+
+      const interceptedAtomCopy: typeof originalAtom = {
+        ...originalAtom,
+        ...('read' in originalAtom && {
           read(get, opts) {
-            return anAtom.read.call(
-              this,
-              (a) => get(getAtom(this, anAtom, a)),
+            return originalAtom.read(
+              (a) => get(getAtom(originalAtom, a)),
               opts,
             );
           },
         }),
-        ...('write' in anAtom && {
+        ...('write' in originalAtom && {
           write(get, set, ...args) {
-            return anAtom.write.call(
-              this,
-              (a) => get(getAtom(this, anAtom, a)),
-              (a, ...v) => set(getAtom(this, anAtom, a), ...v),
+            return originalAtom.write(
+              (a) => get(getAtom(originalAtom, a)),
+              (a, ...v) => set(getAtom(originalAtom, a), ...v),
               ...args,
             );
           },
         }),
+        // eslint-disable-next-line camelcase
+        unstable_is: (a: AnyAtom) => isSelfAtom(a, originalAtom),
```

Comment:
Interesting to see that this doesn't use `this`. Yeah, because it already know originalAtom.